src/distribution.jl

```diff
@@ -0,0 +1,131 @@
+###############################
+# Core Distribution Interface #
+###############################
+
+struct DistributionTrace{T, Dist} <: Trace
+    val::T
+    args
+    score::Float64
+end
+@inline dist(::DistributionTrace{T, Dist}) where {T, Dist} = Dist()
```

Comment:
@georgematheos Unfortunately, this implementation does not support more complicated distributions that are not 'singleton' structs, like the `Mixture` distributions. The problem is that they do not have zero-argument constructors. For example, `HomogeneousMixture`'s definition looks like this:

```
struct HomogeneousMixture{T} <: Distribution{T}
    base_dist::Distribution{T}
    dims::Vector{Int}
end
```

If someone creates such a distribution, and then simulates a trace from it, the trace does not remember the `base_dist` and the `dims`, and so it cannot implement `Gen.get_gen_fn`.

One option is for the `DistributionTrace` to store a reference to the distribution object, and to have `dist(d::DistributionTrace) = d.dist`. That adds slight storage overhead, but maybe not enough to worry about -- it's one more pointer per choice in a trace.

Another option is to say that the Mixtures are not literally subtypes of Distribution -- they are just additional generative functions with ValueChoiceMaps.

I lean toward the first option -- thoughts?

Reply:
> One option is for the DistributionTrace to store a reference to the distribution object, and to have dist(d::DistributionTrace) = d.dist. That adds slight storage overhead, but maybe not enough to worry about -- it's one more pointer per choice in a trace.

I think this may actually add zero storage overhead if the `DistributionTrace` is parametrically typed, and the type in question is a singleton (as it is for `Normal` etc.)! My guess is that Julia optimizes that sort of thing away. But it's worth checking.